records/2022-11-22.security.blazer.md

```diff
@@ -0,0 +1,45 @@
+# Introducing Blazer and its Implementation for Notify
+
+Date: 2022-11-22
+
+## Status
+
+**DRAFT**
+
+## Context
+In order to have a thorough understanding of data in our system, Notify devs wanted a collaborative tool which would allow them to explore the database. The tool would allow us to store queries and share them between developers. An added bonus would be if the tool graphed results in a meaningful manner. The goal of this ADR is to explain the implementation of the tool that was chosen for exploring the database.
+
+## Blazer
+
+Going through a list of products that are available and meet our main requirement of collaboration and graphing we decided on https://github.com/ankane/blazer
+
+### Implementation
+
+Introducing blazer into our system requires us to do the following:
+
+1. Blazer needs access to the main Notify DB
+2. Blazer needs to be secure and auditable
+
+## Deployment
+
+### Blazer Container
+The first step is to add Blazer to the Notify ECR. This way Notify has control over the image that is deployed to our production system. The downside of this is that Notify devs need to maintain this image by pushing security updates
+
+### Database
+Blazer stores all its queries in a Blazer schema under the database. In order to have a sepration of concerns and not use the same database for both Notify data and Blazer data, a new database - "Database Tools" was set up.
+
+### ECS Task
+Set up blazer to run as an ECS Task. Blazer runs as a service within an ECS cluster.
+This meant we didn't have to deploy a Blazer instance inside the EKS cluster and instead separated out the service for its own reliability and security purpose.
+
+### Private Links
+We need to introduce Private Links in order for Blazer to speak to all of Notify's services without exiting the Notify VPC. Previous to the implementation of private links, Blazer as well as our EKS cluster would need to exit the VPC to speak to other AWS cloud services. With the introduction of Blazer/ we were adding a further vector that a third party could attack as Blazer would have to access the internet. We introduced private links into our infrastructure so services can speak to each other without leaving the VPC.
+
+### Diagram Pre Private Links Implementation
+
+![Blazer-preprivate](https://user-images.githubusercontent.com/8869623/203845636-225e0c9e-1f02-4e2e-85bc-4110cbd1d7a2.png)
+
+## Diagram Post Private Links Implementation (Current State of Infrastructure)
+
+![Blazer-postprivate](https://user-images.githubusercontent.com/8869623/203845655-64a67f1a-42d6-4986-88b2-8a5082c6c108.png)
```

Comment:
Let's store the diagram images within the repository itself please.

Reply:
done